src/Rings/msolve/interface.jl

```diff
@@ -0,0 +1,131 @@
+# we take a Singular ideal and extract the following data:
+# an int array lengths storing the lengths of each generator
+# an int array cfs storing the coefficients of each generator
+# an int array exps storing the exponent vectors of each generator
+function convert_singular_ideal_to_array(id::Singular.sideal)
+    ngens   = Singular.ngens(id)
+    nvars   = Singular.nvars(base_ring(id))
+    char    = Singular.characteristic(base_ring(id))
+    lens    = Int32[Singular.length(id[i]) for i in 1:ngens]
+    nterms  = sum(lens)
+    cfs     = Int32[]
+    exps    = Int32[]
+    if char ==  0
+        cfs = BigInt[]
+    else
+        cfs = Int32[]
+    end
+    for i = 1:Singular.ngens(id)
+        if char == 0
+            for c in Singular.coeffs(id[i])
+                push!(cfs, Singular.libSingular.n_GetMPZ(numerator(c).ptr, Singular.ZZ.ptr))
+                push!(cfs, Singular.libSingular.n_GetMPZ(denominator(c).ptr, Singular.ZZ.ptr))
+            end
+        else
+            for c in Singular.coeffs(id[i])
+                push!(cfs, Base.Int(c))
+            end
+        end
+        for e in Singular.exponent_vectors(id[i])
+            for j = 1:nvars
+                push!(exps, Base.Int(e[j]))
+            end
+        end
+    end
+    return lens, cfs, exps
+end
+
+# we know that the terms are already sorted and they are all different
+# coming from GroebnerBasis' F4 computation, so we do not need p_Add_q for
+# the terms, but we can directly set the next pointers of the polynomials
```

Comment:
There is a lot of low level manipulation here that is already in Singular.jl.
Example: https://github.com/oscar-system/Singular.jl/blob/master/test/poly/poly-test.jl#L77
You can even use the same MPolyBuildCtx to construct multply polys (https://github.com/Nemocas/AbstractAlgebra.jl/issues/1043): https://github.com/oscar-system/Singular.jl/blob/master/test/poly/plural-test.jl#L122

I think it would be cleaner and more readable to use these constructors. Also, the MPolyBuildCtx in Singular.jl properly GC protects the pointers as well as sorting the polynomial at the end (is this needed?). However, I do have some questions related to this below.

Reply:
The process of writing the GB in Singular format must be fast. There is no need to sort any terms, they are sorted correctly.  Is there a way how one can disable sorting with these constructors? This step can get crucial once the bases get really big. I will try to play around with the MPolyBuildCtx and see how it behaves.

Reply:
Yeah, the MPolyBuildCtx is not optimal as it has to be safe for the general user. In particular, the final sort cannot be disabled, and the coefficients are copied. If you say sorting is not required, then I would go with what you have here.